JoodHajeer/assignment2/JoodHajeer-2.cpp

```diff
@@ -0,0 +1,252 @@
+	if (!node) 
+		return;
+	if (node == root)
+		node->depth = 0;
+	else
+		node->depth = node->parent->depth + 1;
+	computeDepth(node->left);
+	computeDepth(node->right);
+}
+template <class T>
+bool Binary_Tree<T> ::empty() {
+	return root == nullptr;
+}
+template <class T>
+void Binary_Tree<T> ::clear() {
+	clear(root); 
+	root = nullptr;
+}
+template <class T>
+void Binary_Tree<T> ::clear( BTNode<T>* node) { // post-order traverese to delete nodes in a tree
+	if (!node)
+		return;
+	clear(node->left);
+	clear(node->right);
+	delete node;
+}
+template <class T>
+BTNode<T> * Binary_Tree<T> ::findAvailableParent() {// find first available parent in level order using breadth first traverese
+	if (empty())
+		return nullptr;
+	queue <BTNode<T> *> parents;
+	BTNode<T> * currentNode = root;
+	parents.push(currentNode);
+	while (!parents.empty()) {
+		currentNode = parents.front();
+		parents.pop();
+		if (!currentNode->left  || !currentNode->right) //found an availabe parent
+			return currentNode;
+		parents.push(currentNode->left);
+		parents.push(currentNode->right);
+	}
+}
+template <class T>
+bool Binary_Tree<T> ::find(T &VAL) {//check if value already exists in the tree
+	return find(VAL, root)!=nullptr;
+}
+template <class T>
+BTNode<T> * Binary_Tree<T> ::find(T &VAL, BTNode<T> * node) {//return a pointer to the node with val if exists
+	if (!node)
+		return nullptr;
+	if (node->val == VAL)
+		return node;
+	BTNode<T> * leftSide = find(VAL, node->left);
+	BTNode<T> * rightSide = find(VAL, node->right);
+	if (leftSide)
+		return leftSide;
+	return rightSide;
+}
+template <class T>
+bool Binary_Tree<T> :: insert(T &VAL) {//insert only unique values
+	if (find(VAL))//value already exist in the tree, do not insert it
+		return false;
+	BTNode<T> * prevNode = findAvailableParent();
+	BTNode<T> * newNode = new BTNode<T>(VAL, nullptr, nullptr, prevNode , -1); //val, left, right, patent, depth
+	if (!prevNode) {
+		root = newNode;
+		return true;
+	}
+	newNode->parent = prevNode;
+	if (!prevNode->left)
+		prevNode->left = newNode;
+	else
+		prevNode->right = newNode;
+	return true;
+}
+template <class T>
+void Binary_Tree<T> ::printAncestors(T & VAL) {//print ancestors of node starting from the root
+	BTNode<T> * node = find(VAL, root);
+	if (!node) {//tree is empty or value not found
+		cout << "node was not found in the tree" << endl;
+		return ;
+	}
+	stack <T> ancestorsStack;//save values of ancestors 
+	node = node->parent;
+	while (node) {
+		ancestorsStack.push(node->val);
+		node = node->parent;
+	}
+	if (ancestorsStack.empty()) {
+		cout << "node is the root of the tree" << endl;
+		return;
+	}
+	while (!ancestorsStack.empty()) {
+		T nextAncestor= ancestorsStack.top();
+		ancestorsStack.pop();
+		cout << nextAncestor << " ";
+	}
+	cout << endl;
+}
+template <class T>
+pair<bool,T> Binary_Tree<T> ::commonAncestor(T & VAL1, T & VAL2) { //complexity O(tree height)

```

Comment:
Original way of solving it ^_^

Reply:
Do you prefer if I write the solution we discussed in the meeting?